### ISSUE TITLE
Why do ISpan/ISpanContext not have parity in BaggageItems?

### ISSUE THREAD

**Issue by ndrwrbgs** [opentracing/opentracing-csharp#86]
Might be a question for the specification, but from my reading, the specification is explicit down to this level.

ISpan exposes SetBaggageItem and GetBaggageItem(string), but ISpanContext only allows user to retrieve ALL of the baggage items via `GetBaggageItems()`. Since `ISpan` is supposed to be storing it's baggage in the `ISpanContext`, this seems like an artificial limitation of the API.

**Comment by cwe1ss** [comment]
`ISpanContext` does not have a `SetBaggageItem` method because the [span context should be immutable](https://github.com/opentracing/specification/blob/master/specification.md#spancontext) and the only way to change it is via `ISpan`. 

Not sure if there's a reason for why it doesn't have a `GetBaggageItem(string key)` method though. @yurishkuro @tedsuo do you know of any?

**Comment by yurishkuro** [comment]
is there a critical use case that is not possible with the existing API?  There's generally nothing good about extending the API surface.

**Comment by ndrwrbgs** [comment]
The linked issue #44 has the full context on this @yurishkuro but it's concerning retrieval of information from the `ISpanContext`. Presently, a client must retrieve and iterate through every piece of context to find a specific key:value pair, _unless_ they have access to the `ISpan`. So the use case to support it O(1) lookup of baggage items from the `ISpanContext`.

Although, I think the specification actually mandates the existence of this method as-is today, meaning it wouldn't be a true extension of the API surface, per my understanding.

**Comment by cwe1ss** [comment]
> Presently, a client must retrieve and iterate through every piece of context to find a specific key:value pair, unless they have access to the ISpan.

The only use case I can think of for "instrumentation code" where you have access to a span context but not a span is when you do an `Extract` before you create the span. Is this your use case or do you need it somewhere else? What would be the concrete decision you make based on a baggage item in that case?

> Although, I think the specification actually mandates the existence of this method as-is today, meaning it wouldn't be a true extension of the API surface, per my understanding.

The specification only contains "[Iterate through all baggage items](https://github.com/opentracing/specification/blob/master/specification.md#iterate-through-all-baggage-items)" for SpanContext right now.

**Comment by ndrwrbgs** [comment]
@cwe1ss Good point, I am mixing up my instrumentation work and tracer work - I cannot conceive of a use case where the instrumentation code should be preoccupied with the contents of the Context.

I guess my issue with understanding per the second part is that the spec seems to make SpanContext more of a "because it has to have a name" than a first class thing, and it's the name given to the baggage stored on the Span, which _can_ be accessed by key. Anyway, let's get to my use case to make this simpler:

I have a tracer and it needs the spanID from the parent span when starting a new span. I'd rather not enumerate over the IEnumerable since I've worked in the realm of performance before and that word (`IEnumerable`) is often a hidden red flag.

**Comment by yurishkuro** [comment]
> I have a tracer and it needs the spanID from the parent span when starting a new span. I'd rather not enumerate over the IEnumerable

Tracer typically does not operate on ISpanContext interface, but on the concrete implementation of it, where you have direct access to the fields of the context, including spanID. Also, it's highly unusual to store spanID as baggage, because baggage is the end user concept, something that the application uses, not the tracer.

**Comment by ndrwrbgs** [comment]
I’m not sure that’s the right approach here. Operating on a concrete type is not possible because you should be programming to the interface. Without doing that, you cannot utilize wrapping or - like - anything else that interfaces provide you a benefit of.

Could you help explain why you’d be against exposing index based retrieval of baggage items? It makes no sense to me to be key:value pairs if they can’t be accessed by key - so the fact that we’re having to have a discussion is confusing me a bit — therefore I’m not well equipped to respond to your concerns.

**Comment by yurishkuro** [comment]
> Operating on a concrete type is not possible because you should be programming to the interface. 

The interface is defined for the end user of the API. The concrete tracer always depends on concrete implementation of Span/SpanContext that tracer itself provides. For example, there is no way to ask the Span for which tags and logs it has, yet the tracers need that information to send the spans to the tracing backend. 

> Without doing that, you cannot utilize wrapping

That is simply not true: https://github.com/opentracing-contrib/java-metrics

> Could you help explain why you’d be against exposing index based retrieval of baggage items?

I'm sorry, I don't want to debate some theoretical use case. I've already explained why the API is as it is - to keep it minimal. I've implemented tracers in several languages, as well as written several end-user applications that use baggage, without the need to change the API. So rather than asking "why X?", I suggest you describe what it is that you are trying and not able to achieve with the existing API, a concrete business use case. I have looked through #44, it was closed because it had too many themes going on at once, so I don't know which use case you are referring to.

**Comment by ndrwrbgs** [comment]
I’m confused above, I’m being told both that I should program to the concrete types rather than the interfaces in the tracer, but then also that wrapping is possible (but wrapping would mean I should, as most cases with interfaces, program to the interface). I understand that it’s possible I can just always try to cast the ISpan to MySpan, but as a developer this strikes me very dangerous and brittle, I’d much prefer if either I can be shown my use case is just silly and that’s why it’s not possible with the interfaces, or the interfaces grow to support their conceptual uses. I guess where I’m disagreeing is with ever hard-casting an interface back to a concrete type - doing so makes me feel like the interface wasn’t the right abstraction.

**Comment by ndrwrbgs** [comment]
> The concrete tracer always depends on concrete implementation of Span/SpanContext that tracer itself provides

I think this is where we are disagreeing the most. I understand that it’s possible I can just always try to cast the ISpan to MySpan, but as a developer this strikes me very dangerous and brittle, I’d much prefer if either I can be shown my use case is just silly and that’s why it’s not possible with the interfaces, or the interfaces grow to support their conceptual uses. I guess where I’m disagreeing is with ever hard-casting an interface back to a concrete type - doing so makes me feel like the interface wasn’t the right abstraction.

> For example, there is no way to ask the Span for which tags and logs it has, yet the tracers need that information to send the spans to the tracing backend.

The tracers actually do not need this information stored on the Span itself. The interface does not dictate that these are stored until Finish, and the tracer is responsible for intercepting (and keeping in memory or sending to an external system) the information about the Tag at the time that it is supplied - e.g. via the implementation of the `ISpanBuilder`.

> That is simply not true: https://github.com/opentracing-contrib/java-metrics

It seems like this project has hit the same family of issues I'm hitting, actually, so it rather supports my concerns :( from the project:

> Only works with ActiveSpanSource implementations that don't require a *tracer specific Span implementation*


TL;DR I completely agree with keeping the API minimal, but if I'm struggling to write something as simple as a tracer that logs a metric (same end goal as the project you linked) then it makes me question if the exposed interfaces are sufficient.


So let me take a step back and rephrase a few of the questions that this thread has provoked for me:
1. Is it the official suggestion of OT that tracers NOT program to the interfaces exposed, and on entry into all interface method implementations, cast the supplied interface objects to concrete types?
1. Is it expected in OT that users never access baggage from the `ISpanContext`?
    1. If users are to ever access baggage via the `ISpanContext` is it expected they will _never_ care about the individual items, and only handle them as a collection of items?
1. Is it expected that `ISpanContext` is just (read: mostly) an interface for accessing information typically found on the `ISpan`
    1. (Note - this is how I understood it, and why I was confused about the lack of parity between the parent/source of the data `ISpan` and the accessor interface `ISpanContext`

**Comment by ndrwrbgs** [comment]
P.S. in following links - if https://github.com/opentracing/specification/issues/76 is to ever get taken (seems it has not had any traction though) then all my concerns are negated :)

**Comment by yurishkuro** [comment]
> Is it the official suggestion of OT that tracers NOT program to the interfaces exposed, and on entry into all interface method implementations, cast the supplied interface objects to concrete types?

Yes, although in practice it only applies to SpanContext, since the Span is never passed as argument. The main reason for SpanContext to be used this way is because SpanContext contains implementation specific details that OpenTracing does not care about, even something as simple as traceID is abstracted. Thus there's no way to define an interface that will be suitable for every possible implementation of Tracer and its respective SpanContext.

> Is it expected in OT that users never access baggage from the ISpanContext?

Can't say if it's "expected", but in practice users never interact with SpanConext aside from passing it to tracer methods. 

> Is it expected that ISpanContext is just (read: mostly) an interface for accessing information typically found on the ISpan

Per above, usually you don't need to read anything from it at all, and "typically found on the ISpan" bit is incorrect because there are times when SpanContext exists without a corresponding span, e.g.
```
spanContext := tracer.Extract(...)
newSpan := tracer.StartSpan("name", childOf(spanContext))
```
And that ^^^, btw, is pretty much the full extent of when span context is even in the scope for the end user.

**Comment by yurishkuro** [comment]
As you may have gathered, I am not inclined to debate theory. I've learned long time ago that you don't design software based on theory alone, but by considering specific use cases and trying to generalize from them. In order to keep the solution general and minimal, some use cases end up being harder than others to use.

You gave a very concrete use case, thanks. There is nothing about it that every other OpenTracing implementation in existence doesn't already do. Since the only way to create the context is to use _your tracer_, the solution to the use case is to cast it back from the interface to _your implementation_ of the context. It is the solution that keeps the interface minimal. There is only one use case that I know of that was not possible without having the baggage iterator on the context - and that was a very obscure scenario that, in retrospect, we should've kept out of the API.

**Comment by cwe1ss** [comment]
> @ndrwrbgs I have a tracer I am trying to implement that stores it's relevant information (let's just call it spanid and traceid for ease of communication) in the baggage since that's the only existing place to store them.

Are you talking about your EventHookTracer or is this from a different project?

**Comment by ndrwrbgs** [comment]
@cwe1ss it's different - a real tracer bridging with work's internal tracing system.

I think my take away here, and where I am conceptually in disagreement (but now that I understand where we both are coming from I can at least reconcile that disagreement :smile:) is in the approach of casting the interfaces to a concrete implementation.

@yurishkuro I really have the gut reaction to the desire to keep the interfaces so small that "why then does SpanContext expose BaggageItems at all when it could just be a field-less interface?". I'm struggling to see why we would say these are key:value and not then expose access to them as such - but I guess outside of 'I feel that doesn't make sense', which is too theory based, I'm struggling to find a situation in which someone wouldn't be able to say "just ignore the interface and cast to the concrete type" - which itself makes me feel "why did we bother with the interface?".
^ These are my confusions, but this question is resolved now on how to proceed, so they're not meant to be some demand for change but rather just to explain the sentiment in my confusion. I think a lot of it is coming from the reality that OT came into existence _after_ tracing systems like zipkin and jaeger and it's goal wasn't to be a tracing API/interfaces in the rawest form but rather to provide an access layer that could have a back end of either of those (kind of an implementation -> API rather than an API -> implementation lifecycle).

**Comment by yurishkuro** [comment]
> "why then does SpanContext expose BaggageItems at all when it could just be a field-less interface?"

as I said, it didn't really need that for most cases, aside from one obscure use case that wasn't possible otherwise. That use case needed all baggage items (it didn't know which items are there), and "all" is a smaller API surface than "get by key".

> "why did we bother with the interface?"

That confusion puzzles me - you need an interface to define an API, even if it's an empty interface. And as @cwe1ss mentioned, there is a proposal to extend that API to return implementation-agnostic representation of trace/span IDs.

> kind of an implementation -> API rather than an API -> implementation lifecycle

I wouldn't say that. It is of course true that there were existing implementations at the time of API design, but the goal was to enable vendor-neutral _instrumentation_, i.e. end-user use cases, not to provide an SPI for tracer implementors. Where existing implementations played a role was in them saying "you can't add method X to the API because my tracing system does not have even a concept of X". The API is first and foremost end-user facing, if there is no concrete end-user use case, we try hard not to extend the API for it (tracer implementations can cast and do whatever).

**Comment by ndrwrbgs** [comment]
Thanks for taking the time to reply Yuri! I'll close this issue.
